copilot-chat-copilot.el

```diff
@@ -327,35 +327,93 @@
+  ;;
+  ;;     data: {"choices":[{...,"delta":{"content":"work"}}],...}
+  ;;
+  ;;     data: [DONE]
+  ;;
+  ;; We recieve this piecewise, with this function called with `string' as any substring, completely
+  ;; ignoring the lines and other rules of the protocol. Thus, this function processes line-by-line
+  ;; but needs to be careful to handle partial input any point. We do this by saving a left-over
+  ;; line that failed processing to `copilot-chat-last-data' and reading it on the next call.
+  ;;
+  ;; For instance, this function could be called with three segments like:
+  ;;
+  ;; 1. "data: {...}\n\ndat" (break in the middle of a "data: " prefix)
+  ;; 2. "a: {...}\n\ndata: [D" (break in the middle of some data content)
+  ;; 3. "ONE]\n\n"
+  ;;
+  ;; Those calls will proceed like this:
+  ;;
+  ;; 1. With segment 1, successfully process the first line (`callback' is called with argument "great"), skip
+  ;;    the next empty line, and then fail to process the trailing "dat"; "dat" is saved to
+  ;;    `copilot-chat-last-data'.
+  ;;
+  ;; 2. With segment 2, the value of `copilot-chat-last-data' is first prepended to `string', and
+  ;;    processing continues with "data: {...}\n\ndata: [D". Thus, `callback' is called with "work",
+  ;;    the next line skipped, and then "data: [D" saved to `copilot-chat-last-data'.
+  ;;
+  ;; 3. With segment 3, `copilot-chat-last-data' is prepended to `string', resulting in a value of
+  ;;    "data: [DONE]\n\n". Thus, `callback' is called with the value of `copilot-chat-magic', and
+  ;;    the two trailing empty lines are skipped.
   (when copilot-chat-last-data
     (setq string (concat copilot-chat-last-data string))
     (setq copilot-chat-last-data nil))
 
   (let ((segments (split-string string "\n")))
     (dolist (segment segments)
-      (when (string-prefix-p "data:" segment)
-        (let ((data (substring segment 6)))
-          (if (string= data "[DONE]")
-			  (progn
-				(funcall callback copilot-chat-magic)
-				(setf (copilot-chat-history copilot-chat-instance) (cons (list copilot-chat-curl-answer "assistant") (copilot-chat-history copilot-chat-instance)))
-				(setq copilot-chat-curl-answer nil))
-            (condition-case err
-                (let* ((json-obj (json-parse-string data :object-type 'alist))
-                       (choices (and json-obj (alist-get 'choices json-obj)))
-                       (delta (and (> (length choices) 0) (alist-get 'delta (aref choices 0))))
-                       (token (and delta (alist-get 'content delta))))
-                  (when (and token (not (eq token :null)))
-                    (funcall callback token)
-					(setq copilot-chat-curl-answer (concat copilot-chat-curl-answer token))))
-              (json-parse-error
-				(setq copilot-chat-last-data segment))
-              (json-end-of-file
-               (setq copilot-chat-last-data segment))
-              (error
-               (message (format "erreur : %s" segment))))))))))
```

Comment:
NB. I've removed this error clause. I _think_ any uncaught error will just bubble up and be logged, and can also be debugged by the user with `M-x toggle-debug-on-error` or similar.

Let me know if it is here for a specific reason and I can add it back in to `copilot-chat--extract-segment`.